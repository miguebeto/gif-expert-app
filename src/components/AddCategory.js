import React, { useState } from "react";
import PropsTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  //permite setear o cambiar el value por defecto por el valor ingresado desde el dom en el input cunado ocurra el evento
  const handleInputChange = (e) => {
    setInputValue(e.target.value);

    console.log("handleInputChange llamado");
  };

  const handleSubmit = (e) => {
    //previene que se actualice la pÃ¡gina al ocurrir el evento por defecto
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      //agrega los datos del input como nuevo elemento del arreglo
      setCategories((categories) => [inputValue, ...categories]);
      //vaciamos el inout luego de agregal el valor al arreglo
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p> {inputValue} </p>{" "}
      <input type="text" value={inputValue} onChange={handleInputChange} />{" "}
    </form>
  );
};

//Se utiliza el propTypes and PropsTypes para indicar que el props debe ser obligatorio en el componenete
AddCategory.propTypes = {
  setCategories: PropsTypes.func.isRequired,
};
