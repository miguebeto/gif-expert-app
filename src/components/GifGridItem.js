import React from "react";
import PropTypes from "prop-types";
export const GifGridItem = ({ title, url }) => {
  // console.log(title, url);

  return (
    <div className="card animate__animated animate__pulse">
      {/* desestructuramos las props para mostrar en pantalla la imagen mediante la url y el titulo de cada imagen  */}
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
