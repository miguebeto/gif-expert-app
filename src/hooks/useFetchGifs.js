import React, { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

//Usamos nuestro custom hook al igual que como usamos un componente funcional, agregamos el useState y retornamos lo que que ser치 usado en otros componentes.
export const useFetchGifs = ( category ) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

//no se puede usar async en los efectos porque son sicr칩nicos
    useEffect(()=>{

        //Ejecutamos la promesa para traer la informaci칩n de la petici칩n y guardarla en el estado
        getGifs( category )
        .then(imgs =>{
            //Se mostraran las imagenes luego de 3 segundos
            setTimeout(()=>{      
                setState({
                    data: imgs,
                    loading: false
                })
            },3000)
        })
    },[ category ])

    return state //{data:[], loading: true}
}
