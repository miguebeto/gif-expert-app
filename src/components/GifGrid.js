import React from "react";
import PropTypes from "prop-types";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  //hacemos uso de la propiedad loading dentro del objeto par simular la carga de una pagina.
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      {/* imprimimos en pantalla el nombre de la categoria */}
      <h3>{category}</h3>
      {/* Mostramos cargando cuando el valor de la propiedad loading es true y data cargada cuando pasa a false luego de ejecutarse el setTimeout  */}
      {loading && (
        <p className={"animate__animated animate__flash"}>loading...</p>
      )}
      <div className="card-grid">
        {/* Recorremos el arreglo que obtenemos como resultado extrayendo los valores y pasÃ¡ndolos como props del componente  */}
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
