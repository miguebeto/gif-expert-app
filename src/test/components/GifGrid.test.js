import React from "react";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
import "@testing-library/jest-dom";
//Simula el suo de la peticion fetch
jest.mock("../../hooks/useFetchGifs");

describe("pruebas en el componente <GifGrid />", () => {
  const category = "Evangelion";

  test("debe retornar el componente sin cambios en el rendererizado", () => {
    //muestra lo que debe retornar al hacer cuando el fetch está vacio
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    //trae el componente <GifGrid /> con la props requerida obligatoriamente mediante el propTypes
    const wrapper = shallow(<GifGrid category={category} />);

    //comprueba que el componente se encuentre sin modificaciones en el renderizado
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar items cuando se cargan imágenes useFetchGifs", () => {
    //creamos el objeto que va a simular la respuesta del fetch
    const gifs = [
      {
        id: "ABC",
        url: "https://localhost/cualquier/cosa.jpg",
        title: "cualquier cosa",
      },
      {
        id: "123",
        url: "https://localhost/cualquier/cosa.jpg",
        title: "cualquier cosa",
      },
    ];

    //simulamos el retorno del fetch con la informacion agregada anteriormente
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    //comprueba que el componente se encuentre sin modificaciones en el renderizado
    const wrapper = shallow(<GifGrid category={category} />);

    //comprueba que el componente se encuentre sin modificaciones en el renderizado
    // expect(wrapper).toMatchSnapshot();

    //comprueba que el loading no se encuentra renderizado como un parrafo
    expect(wrapper.find("p").exists()).toBe(false);

    //comprueba que el se pasó el componente <GifGridItem />
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
