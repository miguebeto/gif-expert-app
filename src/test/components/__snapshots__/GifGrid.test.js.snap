// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el componente <GifGrid /> debe mostrar items cuando se cargan im√°genes useFetchGifs 1`] = `
<Fragment>
  <h3>
    Evangelion
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="cualquier cosa"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`pruebas en el componente <GifGrid /> debe retornar el componente sin cambios en el rendererizado 1`] = `
<Fragment>
  <h3>
    Evangelion
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
